### PR TITLE
RegexArrayShapeMatcher - Narrow types in alternation of top level capturing groups

### PR DESCRIPTION
**Pull request by staabm**
fixes the 2 repros of https://github.com/phpstan/phpstan/issues/11221#issuecomment-2198551145

### CODE REVIEW

File: src/Type/Php/RegexCapturingGroup.php [hunk 1]
```diff
@@ -5,49 +5,99 @@
...
 	private function __construct(
 		private ?string $name,
-		private bool $inAlternation,
+		private ?int $alternationId,
 		private bool $inOptionalQuantification,
 		private RegexCapturingGroup|RegexNonCapturingGroup|null $parent,
 	)
 	{
+		$this->id = self::$idGenerator;
+		self::$idGenerator++;
```
Comment (ondrejmirtes) [inline src/Type/Php/RegexCapturingGroup.php]: I don't like global state like that. This should be passed as a plain number from the outside and counted in the instance that creates objects of this class.

File: tests/PHPStan/Analyser/nsrt/preg_match_shapes.php [hunk 2]
```diff
@@ -306,3 +306,21 @@ function groupsOptional(string $size): void
...
 }
+
+function topLevelAlternation(string $size): void
+{
+	if (preg_match('~^(?:(\\d+)x(\\d+)|(\\d+)|x(\\d+))$~', $size, $matches) !== 1) {
+		throw new InvalidArgumentException(sprintf('Invalid size "%s"', $size));
+	}
+	assertType('array{0: string, 1: string, 2: string, 3?: string, 4?: string}', $matches);
+
+	if (preg_match('~^(?:(\\d+)x(\\d+)|(\\d+)|x(\\d+))?$~', $size, $matches) !== 1) {
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/nsrt/preg_match_shapes.php:327]: Better to divide this in multiple functions, reusing the `$matches` variable like that seems dirty to me.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by staabm** [comment]
feedback addressed - thank you

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**